services/job_subscriber.go

```diff
@@ -71,14 +94,69 @@ func (js *JobSubscriber) Disconnect() {
 }
 
 // OnNewHead resumes all pending job runs based on the new head activity.
-func (js *JobSubscriber) OnNewHead(head *models.BlockHeader) {
-	pendingRuns, err := js.Store.JobRunsWithStatus(models.RunStatusPendingConfirmations)
+func (js *jobSubscriber) OnNewHead(head *models.BlockHeader) {
+	pendingRuns, err := js.Store.JobRunsWithStatus(models.RunStatusPendingConfirmations, models.RunStatusInProgress)
 	if err != nil {
 		logger.Error(err.Error())
 	}
+
+	activeJobRunIDs := make(map[string]struct{})
+
+	js.workerMutex.Lock()
+	defer js.workerMutex.Unlock()
 	for _, jr := range pendingRuns {
-		if _, err := ExecuteRunAtBlock(jr, js.Store, jr.Result, head.ToIndexableBlockNumber()); err != nil {
-			logger.Error(err.Error())
+		activeJobRunIDs[jr.ID] = struct{}{}
+
+		workerChannel, ok := js.workers[jr.ID]
+		if !ok {
+			workerChannel = js.createWorkerChannelFor(jr)
```

Comment:
This seems to be our budget version of a worker pool. Shall we encapsulate this into its own type? It could be a light type that wraps a `map[string]struct{}` with this functionality.